src/lib.rs

```diff
@@ -323,42 +379,28 @@ impl DoubleArrayTrie {
         }
     }
 
-    /// Find all matched prefixes. Returns [(end_index, value)].
-    pub fn common_prefix_search(&self, key: &str) -> Option<Vec<(usize, usize)>> {
-        // Reserve the capacity to speed up the performance,
-        // the longest prefix is at most the longest word in the dictionary,
-        // and suppose each iteration matches a prefix,
-        // there should be <longest_word_len> number of prefixes matched.
-        let mut result = Vec::with_capacity(self.longest_word_len);
-
-        let mut b = self.base[0];
-        let mut n;
-        let mut p: usize;
-
-        for (i, c) in key.char_indices() {
-            p = b as usize;
-            n = self.base[p];
-
-            if b == self.check[p] as i32 && n < 0 {
-                result.push((i, (-n - 1) as usize))
-            }
-
-            p = b as usize + c as usize + 1;
-            if b == self.check[p] as i32 {
-                b = self.base[p];
-            } else {
-                return if result.is_empty() { None } else { Some(result) };
-            }
+    /// Iterate thorough all of the matched prefixes. Returning an iterator.
+    pub fn common_prefix_iter<'a>(&'a self, key: &'a str) -> PrefixIter<'a> {
+        let key_len = key.len();
+
+        PrefixIter {
+            key_len: key_len,
+            da: self,
+            char_indices: key.char_indices(),
+            b: self.base[0],
+            p: 0,
+            n: 0,
+            reach_leaf: false,
         }
+    }
 
-        p = b as usize;
-        n = self.base[p];
-
-        if b == self.check[p] as i32 && n < 0 {
-            result.push((key.len(), (-n - 1) as usize));
-            Some(result)
-        } else {
+    /// Find all matched prefixes. Returns [(end_index, value)].
+    pub fn common_prefix_search(&self, key: &str) -> Option<Vec<(usize, usize)>> {
+        let v = self.common_prefix_iter(key).collect::<Vec<(usize, usize)>>();
+        if v.is_empty() {
```

Comment:
``Option<Vec<T>>`` impls `FromIterator<T>`.
replace body with:

    self.common_prefix_iter(key).collect()

Reply:
Performance degradation here?

Reply:
The compiler told me it should be `self.common_prefix_iter(key).map(|x| Some(x)).collect()`. 

I can understand it would be degradation without `size_hint` implemented, `size_hint` would help but not 100%. One way is to allocate at the front and use manual loop to fill the allocated Vec but not by using `collect`. Any alternative better solution?